gh-116146: Add C-API to create module from spec and initfunc

itamaro opened the pull request:
C-API WG decision: https://github.com/capi-workgroup/decisions/issues/77

<!-- gh-issue-number: gh-116146 -->
* Issue: gh-116146
<!-- /gh-issue-number -->

itamaro committed:
gh-116146: Add new C-API to create builtin from spec and initfunc

encukou commented on Python/import.c:
```diff
@@ -2322,8 +2322,27 @@ is_builtin(PyObject *name)
+
 static PyObject*
-create_builtin(PyThreadState *tstate, PyObject *name, PyObject *spec)
+create_builtin_ex(
```
This is `static`; you can add an argument to it and update all callers.

itamaro replied:
sorry about the force push, but I already had the branch rebased locally before getting comments on the draft PR.
I updated all the callers of `create_builtin` as suggested!

encukou reviewed (commented):
This looks reasonable.

I'm proposing a replacement for `Init` functions in PEP-793, but that PEP adds `PyModule_FromSlotsAndSpec` and `PyModule_Exec` which should work for similar use cases.
Would it make sense to switch the *spec* & *initfunc* arguments to match that proposal?

I'm not sure if the word `Builtin` in the name is necessary -- is that to match the internal `create_builtin`? Would something like `PyImport_LoadModuleFromInitfuncAndSpec` work better?

Do you want to submit this to the [C API WG](https://github.com/capi-workgroup/decisions/issues)?

vstinner commented on Python/import.c:
```diff
@@ -2397,6 +2407,35 @@ create_builtin(PyThreadState *tstate, PyObject *name, PyObject *spec)
+    if (!PyUnicode_Check(name)) {
+        PyErr_Format(PyExc_TypeError,
+                     "name must be string, not %.200s",
+                     Py_TYPE(name)->tp_name);
```
```suggestion
                     "spec name must be string, not %T", name);
```

itamaro replied:
sorry about the force push, but I already had the branch rebased locally before getting comments on the draft PR.
I applied this suggestion.

itamaro committed:
Apply review feedback and CAPI WG decision

- use `PyImport_CreateModuleFromInitfunc` for the new API
- no need to introduce `create_builtin_ex` - update `create_builtin` and all callers
- update exception message

itamaro committed:
Add tests for the PyImport_CreateModuleFromInitfunc API

itamaro committed:
Document the new `PyImport_CreateModuleFromInitfunc` C-API

itamaro committed:
Add news entry

itamaro committed:
Fix lint

itamaro committed:
fix CI

- remove unused `found` variable
- use `my_test_extension` instead of `embedded_ext` (the former is free-threading-ready, the latter prints a warning)

kumaraditya303 commented on Python/import.c:
```diff
@@ -2438,6 +2443,28 @@ create_builtin(PyThreadState *tstate, PyObject *name, PyObject *spec)
+{
+    PyThreadState *tstate = _PyThreadState_GET();
+
+    PyObject *name = PyObject_GetAttrString(spec, "name");
```
```suggestion
    PyObject *name = PyObject_GetAttr(spec, &_Py_ID(name));
```

itamaro committed:
Apply Kumar's review suggestion

Co-authored-by: Kumar Aditya <kumaraditya@python.org>

encukou committed:
PyImport_CreateModuleFromInitfunc: Doc & test update (#27)

* Test single-phase init as well; don't use private APIs in test
* Doc update

---------

Co-authored-by: Itamar Oren <itamarost@gmail.com>

vstinner commented on Doc/c-api/import.rst:
```diff
@@ -333,3 +333,24 @@ Importing Modules
+
+   On error, return NULL with an exception set.
+
+   .. versionadded:: 3.15
```
```suggestion
   .. versionadded:: next
```

vstinner commented on Include/cpython/import.h:
```diff
@@ -10,6 +10,15 @@ struct _inittab {
+PyImport_CreateModuleFromInitfunc(
+    PyObject *spec,
+    PyObject* (*initfunc)(void)
+    );
```
```suggestion
PyAPI_FUNC(PyObject *) PyImport_CreateModuleFromInitfunc(
    PyObject *spec,
    PyObject* (*initfunc)(void));
```

vstinner commented on Programs/_testembed.c:
```diff
@@ -2213,6 +2223,101 @@ static int test_repeated_init_and_inittab(void)
     return 0;
 }
 
+static PyObject* create_module(PyObject* self, PyObject* spec) {
```
```suggestion
static PyObject*
create_module(PyObject* self, PyObject* spec)
{
```

vstinner commented on Programs/_testembed.c:
```diff
@@ -2213,6 +2223,101 @@ static int test_repeated_init_and_inittab(void)
+    return NULL;
+}
+
+static PyObject* exec_module(PyObject* self, PyObject* mod) {
```
```suggestion
static PyObject*
exec_module(PyObject* self, PyObject* mod)
{
```

vstinner commented on Programs/_testembed.c:
```diff
@@ -2213,6 +2223,101 @@ static int test_repeated_init_and_inittab(void)
+    return PyModuleDef_Init(&create_static_module_def);
+}
+
+static int test_create_module_from_initfunc(void)
```
```suggestion
static int
test_create_module_from_initfunc(void)
```

vstinner commented on Python/import.c:
```diff
@@ -2438,6 +2443,28 @@ create_builtin(PyThreadState *tstate, PyObject *name, PyObject *spec)
+    PyObject *spec, PyObject* (*initfunc)(void))
+{
+    PyThreadState *tstate = _PyThreadState_GET();
+
```
I would suggest setting a ValueError if initfunc is `NULL`, or maybe a `SystemError` (call `_PyErr_BadInternalCall()`).

itamaro committed:
Apply Victor's review suggestions

Co-authored-by: Victor Stinner <vstinner@python.org>

itamaro committed:
Set error if initfunc is NULL

itamaro committed:
Account for the RuntimeWarning when import a singlephase init extension under the free-threaded build

vstinner commented on Misc/NEWS.d/next/C_API/2025-11-08-10-51-50.gh-issue-116146.pCmx6L.rst:
```diff
@@ -0,0 +1,2 @@
+Add a new :c:func:`PyImport_CreateModuleFromInitfunc` C-API for creating a
+module from a **spec** and **initfunc**. Patch by Itamar Oren.
```
```suggestion
module from a *spec* and *initfunc*. Patch by Itamar Oren.
```

vstinner reviewed (commented):
The new function `PyImport_CreateModuleFromInitfunc()` should be documented in `Doc/whatsnew/3.15.rst`.

encukou commented on Doc/c-api/import.rst:
```diff
@@ -333,3 +333,24 @@ Importing Modules
+   the same as for :c:func:`PyImport_AppendInittab`.
+
+   On success, create and return a module object.
+   This module will not be initialized; call :c:func:`PyModule_Exec`
```
```suggestion
   This module will not be initialized; call :c:func:`!PyModule_Exec`
```
When I merge `PyModule_Exec`'s docs, I'll take the bang out.

encukou replied:
(Or if you'd like to review #141197, I'd appreciate it :)

kumaraditya303 commented on Include/cpython/import.h:
```diff
@@ -10,6 +10,13 @@ struct _inittab {
+// without needing to go through inittab
+PyAPI_FUNC(PyObject *) PyImport_CreateModuleFromInitfunc(
+    PyObject *spec,
+    PyObject* (*initfunc)(void));
```
```suggestion
    PyObject *(*initfunc)(void));
```

kumaraditya303 commented on Programs/_testembed.c:
```diff
@@ -166,6 +166,8 @@ static PyModuleDef embedded_ext = {
 static PyObject*
 PyInit_embedded_ext(void)
 {
+    // keep this a single-phase initialization module;
```
```suggestion
    // keep this as a single-phase initialization module;
```

kumaraditya303 commented on Python/import.c:
```diff
@@ -2438,6 +2443,33 @@ create_builtin(PyThreadState *tstate, PyObject *name, PyObject *spec)
 
+PyObject*
+PyImport_CreateModuleFromInitfunc(
+    PyObject *spec, PyObject* (*initfunc)(void))
```
```suggestion
    PyObject *spec, PyObject *(*initfunc)(void))
```

itamaro committed:
Apply suggestions from code review

Co-authored-by: Petr Viktorin <encukou@gmail.com>
Co-authored-by: Kumar Aditya <kumaraditya@python.org>
Co-authored-by: Victor Stinner <vstinner@python.org>

itamaro committed:
Mention new C-API in whatsnew

vstinner reviewed (approved):
LGTM. Thanks for the updates!

itamaro committed:
Merge branch 'main' into gh-116146-create-builtin-initfunc
